Proposal on consistent approach for running GitHub actions (#3961)
alphagov/govuk-developer-docs: source/manual/test-and-build-a-project-with-github-actions.html.md

### When the CI workflow should run

The workflow should be configured to run when there is a [`push`][push-event]
to the default branch (typically "main") and when a

> [`pull_request`][pull-request-event] opened. This will mean that there is a
> CI run when a branch is merged (the `push` to main) and it will run against

Reviewer: ```suggestion
[`pull_request`][pull-request-event] opened. This means CI runs when a branch is merged (the `push` to main) and it runs against
```

in a pull request. We prefer to run against a `pull_request` event to all
`push` events, as this allows external contributors to have pull requests
tested and because `pull_request` merges the changes with

> the repository default branch.

Reviewer 2: I don't understand this last bit. Don't GitHub actions run against the branch as-is, whether or not the trigger was `push` or `pull_request`?

Reviewer: So I only found out about the details of this recently myself (thanks to @kevindew for putting me onto it) — pull_request uses the `refs/pull/<pr_number>/merge` reference that GitHub uses for setting up the merge. (I believe this is essentially the PR branch rebased onto the destination branch, but don't quote me on that.) Compare https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request with https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#push

If the PR branch is up-to-date with main then there should be no practical difference, but if they've diverged significantly and "Require branches to be up-to-date before merging" isn't enabled then it could mean you're testing against an old version of the rest of the codebase.

Reviewer 2: Thanks for the explanation! Perhaps we should amend that sentence to link back to your comment? 🙏

Example configuration:

```yml
# ./github/workflows/ci.yml
on:
  push:
    branches:
      - main
  pull_request:

```

> To have CI runs be able to run outside of a pull request, for example before a
> pull request is opened, you may configure a repository to have a
> `workflow_dispatch` event so it can be ran from the GitHub Actions user

Reviewer: ```suggestion
To run CI on-demand or outside of a pull request, for example before a
pull request is opened, you may configure a repository to have a
`workflow_dispatch` event so you can run it from the GitHub Actions user
```

interface. If you do this you will need to configure the checkout action to
reference the appropriate commit.

Example configuration:

```yml
on:
  push:
    branches:
      - main
  pull_request:
  workflow_dispatch:
    inputs:
      ref:
        description: 'The branch, tag or SHA to checkout'
        required: false
        type: string

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
        with:
          ref: ${{ inputs.ref || github.ref }}
```

We previously recommended CI jobs to run on `[push, pull_request]` as per

> [RFC-123][] however this caused an appearance of duplicate runs of a CI job

Reviewer: It definitely runs real duplicate jobs, unfortunately.
```suggestion
[RFC-123][] however this caused duplicate runs of a CI job
```

Reviewer 2: Worth updating the PR description too, as it still references "appearance of", @kevindew

of a pull request (one for the `push` another for the `pull_request`)

> which were both superfluous for our testing and a point of confusion.

Reviewer: ```suggestion
which were both superfluous for our testing and a point of confusion, as well as wasting resources and in some cases exacerbating quota issues caused by expensive tests.
```

Author: I've put a more succinct, and slightly softer, phrasing in.
